File: LeetCode_Java/binarytree/AllNodesDistanceK.java
package binarytree;

import java.util.*;

public class AllNodesDistanceK {

    public List<Integer> distanceK(TreeNode root, TreeNode target, int k) {

        Map<TreeNode, TreeNode> map = new HashMap<>();
        buildMap(root, null, map);

        List<Integer> result = new ArrayList<>();

        Queue<TreeNode> queue = new LinkedList<>();
        Set<TreeNode> visited = new HashSet<>();
        queue.offer(target);
        visited.add(target);

        int count = 0;

        while (!queue.isEmpty()) {
            if (count == k) {
                for (TreeNode node : queue) {
                    result.add(node.val);
                }
                break;
            }

            int size = queue.size();
            for (int i = 0; i < size; i++) {
                TreeNode cur = queue.poll();

                if (cur.left != null && !visited.contains(cur.left)) {
                    queue.offer(cur.left);
                    visited.add(cur.left);
                }

                if (cur.right != null && !visited.contains(cur.right)) {
                    queue.offer(cur.right);
                    visited.add(cur.right);
                }

                TreeNode parent = map.get(cur);
                if (parent != null && !visited.contains(parent)) {
                    queue.offer(parent);
                    visited.add(parent);
                }
            }

            count++;
        }

        return result;
    }

    private void buildMap(TreeNode node, TreeNode parent, Map<TreeNode, TreeNode> map) {
        if (node == null) {
            return;
        }

        map.put(node, parent);
        buildMap(node.left, node, map);
        buildMap(node.right, node, map);
    }

    public static void main(String[] args) {
        AllNodesDistanceK solution = new AllNodesDistanceK();

        TreeNode root = new TreeNode(3);
        root.left = new TreeNode(5);
        root.right = new TreeNode(1);
        root.left.left = new TreeNode(6);
        root.left.right = new TreeNode(2);
        root.left.right.left = new TreeNode(7);
        root.left.right.right = new TreeNode(4);
        root.right.left = new TreeNode(0);
        root.right.right = new TreeNode(8);

        TreeNode target = root.left;  // target node is 5
        int K = 2;
        List<Integer> result1 = solution.distanceK(root, target, K);
        System.out.println(result1);  // Expected output: [7, 4, 1]

        target = root.left.right.right;  // target node is 4
        K = 1;
        List<Integer> result2 = solution.distanceK(root, target, K);
        System.out.println(result2);  // Expected output: [2]
    }
}

File: LeetCode_Python/binarytree/AllNodesDistanceK.py
from TreeNode import TreeNode
from collections import defaultdict
from collections import deque


class AllNodesDistanceK:
    def distanceK(self, root: TreeNode, target: TreeNode, k: int) -> list[int]:
        self.graph = defaultdict(list)
        self.buildGraph(root, None, self.graph)

        queue = deque([(target, 0)])
        visited = set([target])
        res = []

        while queue:
            node, distance = queue.popleft()
            if distance == k:
                res.append(node.val)
                continue
            for neighbor in self.graph[node]:
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append((neighbor, distance + 1))
        return res

    def distanceKByDFS(self, root: TreeNode, target: TreeNode, k: int) -> list[int]:
        self.graph = defaultdict(list)
        self.buildGraph(root, None, self.graph)

        visited = set([target])
        res = []

        def dfs(node: TreeNode, distance: int) -> None:
            if distance == k:
                res.append(node.val)
                return
            for nei in self.graph[node]:
                if nei not in visited:
                    visited.add(nei)
                    dfs(nei, distance + 1)

        dfs(target, 0)
        return res

    def buildGraph(self, node: TreeNode, parent: TreeNode, graph: defaultdict) -> None:
        if node and parent:
            graph[node].append(parent)
            graph[parent].append(node)
        if node.left:
            self.buildGraph(node.left, node, graph)
        if node.right:
            self.buildGraph(node.right, node, graph)


# Time Complexity: O(N)
# Space Complexity: O(N)

if __name__ == "__main__":
    solution = AllNodesDistanceK()

    root = TreeNode(3)
    root.left = TreeNode(5)
    root.right = TreeNode(1)
    root.left.left = TreeNode(6)
    root.left.right = TreeNode(2)
    root.right.left = TreeNode(0)
    root.right.right = TreeNode(8)
    root.left.right.left = TreeNode(7)
    root.left.right.right = TreeNode(4)

    target = root.left
    k = 2
    print("Expected Output: [1, 7, 4]")
    print("Actual Output:", solution.distanceK(root, target, k))
    print("Actual Output:", solution.distanceKByDFS(root, target, k))
